File: android/src/com/ocproducts/composr/sdk/utilities/CMS_Arrays.java
package com.ocproducts.composr.sdk.utilities;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Set;

/*

CMS_Arrays

array collapse_1d_complexity(string key, array arr) - takes a list of maps and turns it into a simple list by extracting just one element from each map
map collapse_2d_complexity(string keyKey, string valKey, array arr) - takes a list of maps and turns it into a simple map by extracting just two elements from each map
array explode(string sep, string str) - see the PHP manual
string implode(string sep, array arr) - see the PHP manual

*/
@SuppressWarnings({"unchecked","rawtypes"})
public class CMS_Arrays {
	
	public static List<String> collapse_1d_complexity(String key, List<Map<String, String>> arr) {
		ArrayList<String> array = new ArrayList<String>();
		for (Map<String,String> hashMap : arr) {
			String val = hashMap.get(key);
			if (val != null) {
				array.add(val);
			}
		}
		return array;
	}
	
	public static Map<String, String> collapse_2d_complexity(String keyKey, String valKey, List<Map<String, String>> arr) {
		HashMap<String, String> map = new HashMap<String, String>();
		for (Map<String,String> hashMap : arr) {
			String key = hashMap.get(keyKey);
			String val = hashMap.get(valKey);
			if (key != null && val != null) {
				map.put(key, val);
			}
		}
		return map;
	}
	
	public static List<String> explode(String sep, String str) {
		return Arrays.asList(str.split(sep));
	}
	
	public static String implode(String sep, List<String> arr) {
		StringBuilder sb = new StringBuilder();
	    Iterator<?> iter = arr.iterator();
	    if (iter.hasNext())
	        sb.append(iter.next().toString());
	    while (iter.hasNext()) {
	        sb.append(sep);
	        sb.append(iter.next().toString());
	    }
	    return sb.toString();
	}
	
	public static String mapToString(Map<String, String> map) {
		if (map == null || map.isEmpty()) {
			return "";
		}
		Set<String> keySet = map.keySet();
		if (keySet.isEmpty()) {
			return "";
		}
		ArrayList<String> mapElements = new ArrayList<String>();
		for (String key : map.keySet()) {
			mapElements.add(key + ":" + map.get(key));
		}
		return CMS_Arrays.implode(";", mapElements);
	}
	
	public static Map<String, String> stringToMap(String mapString) {
		List<String> mapElements = CMS_Arrays.explode(";", mapString);
		HashMap<String, String> map = new HashMap<String, String>();
		for (String string : mapElements) {
			String[] keyValuePair = string.split(":");
			if (keyValuePair != null && keyValuePair.length < 2) {
				continue;
			}
			map.put(keyValuePair[0], keyValuePair[1]);
		}
		return map;
	}
	
	public static boolean in_array(String key, List<String> array) {
		if(array!=null && array.size()>0) {
			for(String string:array) {
				if(string.equals(key)) {
					return true;
				}
			}
		}
		return false;
	}
	
	public static boolean in_array(Float key, List<Float> array) {
		if(array!=null && array.size()>0) {
			for(Float floatValue:array) {
				if(floatValue.equals(key)) {
					return true;
				}
			}
		}
		return false;
	}
	
	public static boolean in_array(Integer key, List<Integer> array) {
		if(array!=null && array.size()>0) {
			for(Integer integer:array) {
				if(integer.equals(key)) {
					return true;
				}
			}
		}
		return false;
	}
	
	public static <T> List<T> array_merge(List<T> array1, List<T> array2) {
		if(array1 !=null && array1.size()>0 && array2!=null && array2.size()>0) {
			array1.addAll(array2);
			return array1;
		}
		if (array1 == null) {
			return array2;
		}
		if (array2 == null) {
			return array1;
		}
		return null;
	}
	
	public static <T> boolean array_key_exists(String key, Map<String, T> map) {
		if(map!=null && map.size()>0){
			for(String keyValue:map.keySet()) {
				if(keyValue!=null && key.equals(keyValue)) {
					return true;
				}
			}
		}
		return false;	
	}
	
	public static <T> List<T> array_unique(List<T> array) {
		Set<T> set=new HashSet<T>();
		List<T> resultList=new ArrayList<T>();
		if(array!=null && array.size()>0) {
			set.addAll(array);
		}
		resultList.addAll(set);
		return resultList;
	}
	
	public static void sort_maps_by(List<Map<String,String>> array, final String key) {
		Collections.sort(array, new Comparator<Map<String,String>>() {

			@Override
			public int compare(Map<String,String> lhs, Map<String,String> rhs) {
				return lhs.get(key).compareTo(rhs.get(key));
			}
		});
	}
	
	public static void sort(List array) {
		Collections.sort(array);
	}
	
	public static int count(List array) {
		if(array!=null) {
			return array.size();
		}
		return 0;
	}
	
	public static int count(Map map) {
		if(map!=null) {
			return map.size();
		}
		return 0;
	}
	
	public static <T> Map<T, Map<String, T>> list_to_map(String key, List<Map<String, T>> inputList) {
		Map<T, Map<String, T>> resultDict = new HashMap<T, Map<String,T>>();
		for (int i = 0; i < inputList.size(); i++) {
			resultDict.put(inputList.get(i).get(key), inputList.get(i));
		}
		return resultDict;
	}
	
	public static <T, U> List<?> Array_values(boolean useKey, Map<T, U> dict) {
		if (useKey) {
			return new ArrayList<T>(dict.keySet());
		}
		List<U> values = new ArrayList<U>();
		for (T key : dict.keySet()) {
			values.add(dict.get(key));
		}
		return values;
	}
}

File: iOS/CMS_SDK/Composr Mobile SDK/CMS_Utilities/CMS_Arrays.m
//
//  CMS_Arrays.m
//  Composr Mobile SDK
//
//  Created by Aaswini on 06/08/14.
//  Copyright (c) 2014 Aaswini. All rights reserved.
//

#import "CMS_Arrays.h"
#import "CMS_Strings.h"

@implementation CMS_Arrays

/**
 *  Takes a list of maps and turns it into a simple list by extracting just one element from each map
 *
 *  @param key key for which value needs to be fetched
 *  @param arr source array containing many dictionaries
 *
 *  @return array of values corresponding to the specified key from each dictionary
 */
+ (NSArray *)collapse_1d_complexity:(NSString *)key :(NSArray *)arr{
    return [arr valueForKey:key];
}

/**
 *  Takes a list of maps and turns it into a simple dictionary by extracting just two elements from each dictionary.
 *
 *  @param keyKey key for which value to be fetched. This value is kept as key in the returning dictionary.
 *  @param valKey key for which value to be fetched to be kept as value for the first parameter in returning dictionary
 *  @param arr    source array of dictionaries
 *
 *  @return dictionary with keyvalue pair for the supplied keys from all dictionaries in the source array
 */
+ (NSDictionary *)collapse_2d_complexity:(NSString *)keyKey :(NSString *)valKey :(NSArray *)arr{
    NSMutableDictionary *result = [NSMutableDictionary new];
    for (NSDictionary *dict in arr) {
        NSString *key = [dict valueForKey:keyKey];
        NSString *value = [dict valueForKey:valKey];
        [result setObject:value forKey:key];
    }
    return result;
}

/**
 *  Returns an array by splitting the string with the supplied seperator
 *
 *  @param sep separator
 *  @param str source string
 *
 *  @return array of components in string by splitting with separator
 */
+ (NSArray *)explode:(NSString *)sep :(NSString *)str{
    return [str componentsSeparatedByString:sep];
}

/**
 *  Returns a string by joining the supplied array with the supplied seperator
 *
 *  @param sep separator
 *  @param arr source array
 *
 *  @return returns string by joining the components in the array with the separator
 */
+ (NSString *)implode:(NSString *)sep :(NSArray *)arr{
    return [arr componentsJoinedByString:sep];
}

/**
 *  Returns array of keys or values from a dictionary
 *
 *  @param useKey YES - returns array of keys, NO- returns array of values
 *  @param dict   source dictionary
 *
 *  @return returns array of keys or values from the dictionary
 */
+ (NSArray *)Array_values:(BOOL)useKey :(NSDictionary *)dict{
    NSMutableArray *result = [NSMutableArray new];
    for (NSString *key in dict.allKeys) {
        if (useKey) {
            [result addObject:key];
        }
        else{
            [result addObject:[dict valueForKey:key]];
        }
    }
    return result;
}

/**
 *  Returns if the array contains the given key
 *
 *  @param key   value to check for
 *  @param array source array
 *
 *  @return YES if array contains key else NO
 */
+ (BOOL)in_array:(id)key :(NSArray *)array {
    return [array containsObject:key];
}

/**
 *  Merges two arrays to a single array
 *
 *  @param array1 source array 1
 *  @param array2 source array 2
 *
 *  @return single array containing values of array1 and array2
 */
+ (NSArray *)array_merge:(NSArray *)array1 :(NSArray *)array2 {
    return [array1 arrayByAddingObjectsFromArray:array2];
}

/**
 *  Check of given map contains the given key
 *
 *  @param key String key to check for
 *  @param map source map
 *
 *  @return YES if map contains the key else NO
 */
+ (BOOL)array_key_exists:(id)key :(NSDictionary *)map {
    return [[map allKeys] containsObject:key];
}

/**
 *  Removes duplicate values from array
 *
 *  @param array source array
 *
 *  @return Returns array by removing duplicate values
 */
+ (NSArray *)array_unique:(NSArray *)array {
    return [[NSOrderedSet orderedSetWithArray:array] array];
}

/**
 *  Sorts the given array of maps based on value of a given key
 *
 *  @param array source array of maps
 *  @param key   key to sort the maps on
 */
+ (void)sort_maps_by:(NSArray **)array :(NSString *)key {
    *array = [*array sortedArrayUsingComparator:^NSComparisonResult(NSDictionary* obj1, NSDictionary* obj2) {
        if (![obj1 isKindOfClass:[obj2 class]]) {
            return [[NSString stringWithFormat:@"%@",[obj1 objectForKey:key]] compare:[NSString stringWithFormat:@"%@",[obj2 objectForKey:key]] options:NSCaseInsensitiveSearch];
        }
        else if ([[obj1 objectForKey:key] isKindOfClass:[NSNumber class]]) {
            return [[obj1 objectForKey:key] compare:[obj2 objectForKey:key]];
        }
        return [[obj1 objectForKey:key] compare:[obj2 objectForKey:key] options:NSCaseInsensitiveSearch];
    }];
}

/**
 *  Sort array of strings or numbers
 *
 *  @param array source array of strings or numbers
 */
+ (void)sort:(id*)array {
    *array = [*array sortedArrayUsingComparator:^NSComparisonResult(id obj1, id obj2) {
        if (![obj1 isKindOfClass:[obj2 class]]) {
            return [[NSString stringWithFormat:@"%@",obj1] compare:[NSString stringWithFormat:@"%@",obj2] options:NSCaseInsensitiveSearch];
        }
        else if([obj1 isKindOfClass:[NSNumber class]]) {
            return [obj1 compare:obj2];
        }
        return [obj1 compare:obj2 options:NSCaseInsensitiveSearch];
    }];
}

/**
 *  Returns number of elements in an array or map
 *
 *  @param array source array or map
 *
 *  @return number of elements in array or map
 */
+ (int)count:(id)array {
    if ([array isKindOfClass:[NSArray class]]) {
        return (int)[(NSArray *)array count];
    }
    else if ([array isKindOfClass:[NSDictionary class]]) {
        return (int)[(NSDictionary *)array count];
    }
    else if ([array isKindOfClass:[NSString class]]) {
        return (int)[CMS_Strings strlen:array];
    }
    return 0;
}

/**
 *  Returns map corresponding to a key from array of maps
 *
 *  @param key  key
 *  @param dict array of maps
 *
 *  @return map
 */
+ (NSDictionary *) list_to_map:(NSString *)key :(NSArray *)dict{
    NSDictionary *ResultDict=[[NSDictionary alloc] init];
    ResultDict=[NSDictionary dictionaryWithObjects:dict forKeys:[dict valueForKey:key]];
    return ResultDict;
}

@end
